File: program/src/state.rs
use crate::{error::JabberError, utils::order_keys};
use bonfida_utils::BorshSize;
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{
    account_info::AccountInfo, clock::UnixTimestamp, entrypoint::ProgramResult,
    program_error::ProgramError, pubkey::Pubkey,
};

pub const MAX_NAME_LENGTH: usize = 100;
pub const MAX_BIO_LENGTH: usize = 100;
pub const MAX_GROUP_NAME_LEN: usize = 100;
pub const MAX_ADMIN_LEN: usize = 10;
pub const MAX_HASH_LEN: usize = 64;

pub const MAX_PROFILE_LEN: usize =
    1 + MAX_HASH_LEN + MAX_NAME_LENGTH + MAX_BIO_LENGTH + 8 + 1 + 4 + 4;

pub const MAX_GROUP_THREAD_LEN: usize = 1 // tag
    + 1 // visible
    + 32 // owner
    + 8 // last message time
    + 32 // destination_wallet
    + 4 // msg_count
    + 8 // lamports_per_message
    + 1 // bump
    + 1 // media_enabled
    + 1 // admin_only
    + (4 + MAX_HASH_LEN) // group_pic_hash
    + (4 + MAX_GROUP_NAME_LEN) // group_name
    + (4 + MAX_ADMIN_LEN * 32); // admins

pub const MAX_GROUP_THREAD_INDEX: usize = 1 + 4 + MAX_GROUP_NAME_LEN + 32 + 32;

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone, Copy, BorshSize)]
pub enum Tag {
    Uninitialized,
    Profile,
    Thread,
    Message,
    Jabber,
    GroupThread,
    GroupThreadIndex,
    Subscription,
}

#[derive(BorshSerialize, BorshDeserialize)]
pub struct Profile {
    pub tag: Tag,
    pub picture_hash: String,
    pub display_domain_name: String,
    pub bio: String,
    pub lamports_per_message: u64,
    pub bump: u8,
    pub tips_sent: u32,
    pub tips_received: u32,
}

impl Profile {
    pub const SEED: &'static str = "profile";

    pub fn find_from_user_key(user_key: &Pubkey, program_id: &Pubkey) -> (Pubkey, u8) {
        let (user_profile_key, bump) = Pubkey::find_program_address(
            &[Profile::SEED.as_bytes(), &user_key.to_bytes()],
            program_id,
        );
        (user_profile_key, bump)
    }

    pub fn create_from_keys(user_key: &Pubkey, program_id: &Pubkey, bump: u8) -> Pubkey {
        let seeds = &[Profile::SEED.as_bytes(), &user_key.to_bytes(), &[bump]];
        Pubkey::create_program_address(seeds, program_id).unwrap()
    }

    pub fn new(
        picture_hash: String,
        display_domain_name: String,
        bio: String,
        lamports_per_message: u64,
        bump: u8,
    ) -> Self {
        Self {
            tag: Tag::Profile,
            display_domain_name,
            picture_hash,
            bio,
            lamports_per_message,
            bump,
            tips_sent: 0,
            tips_received: 0,
        }
    }

    pub fn save(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    pub fn from_account_info(a: &AccountInfo) -> Result<Profile, ProgramError> {
        let mut data = &a.data.borrow() as &[u8];
        if data[0] != Tag::Profile as u8 && data[0] != Tag::Uninitialized as u8 {
            return Err(JabberError::DataTypeMismatch.into());
        }
        let result = Profile::deserialize(&mut data)?;
        Ok(result)
    }
}

#[derive(BorshSerialize, BorshDeserialize, BorshSize)]
pub struct Thread {
    pub tag: Tag,
    pub msg_count: u32,
    pub user_1: Pubkey,
    pub user_2: Pubkey,
    pub last_message_time: UnixTimestamp,
    pub bump: u8,
}

impl Thread {
    pub const SEED: &'static str = "thread";

    pub fn new(user_1: Pubkey, user_2: Pubkey, bump: u8, last_message_time: UnixTimestamp) -> Self {
        Self {
            tag: Tag::Thread,
            msg_count: 0,
            user_1,
            user_2,
            bump,
            last_message_time,
        }
    }

    pub fn find_from_users_keys(
        user_1: &Pubkey,
        user_2: &Pubkey,
        program_id: &Pubkey,
    ) -> (Pubkey, u8) {
        let (key_1, key_2) = order_keys(user_1, user_2);
        let (thread_key, bump) = Pubkey::find_program_address(
            &[
                Thread::SEED.as_bytes(),
                &key_1.to_bytes(),
                &key_2.to_bytes(),
            ],
            program_id,
        );
        (thread_key, bump)
    }

    pub fn create_from_user_keys(
        user_1: &Pubkey,
        user_2: &Pubkey,
        program_id: &Pubkey,
        bump: u8,
    ) -> Pubkey {
        let (key_1, key_2) = order_keys(user_1, user_2);
        let seeds = &[
            Thread::SEED.as_bytes(),
            &key_1.to_bytes(),
            &key_2.to_bytes(),
            &[bump],
        ];
        Pubkey::create_program_address(seeds, program_id).unwrap()
    }

    pub fn save(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    pub fn from_account_info(a: &AccountInfo) -> Result<Thread, ProgramError> {
        let mut data = &a.data.borrow() as &[u8];
        if data[0] != Tag::Thread as u8 && data[0] != Tag::Uninitialized as u8 {
            return Err(JabberError::DataTypeMismatch.into());
        }
        let result = Thread::deserialize(&mut data)?;
        Ok(result)
    }

    pub fn increment_msg_count(&mut self, current_time: i64) {
        self.last_message_time = current_time;
        self.msg_count += 1;
    }
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, BorshSize)]
pub enum MessageType {
    EncryptedText,
    UnencryptedText,
    EncryptedMedia,
    UnencryptedMedia,
    Deleted,
}

#[derive(BorshDeserialize, BorshSerialize, BorshSize)]
pub struct Message {
    pub tag: Tag,
    // Message type
    pub kind: MessageType,
    // Time at which the message was sent
    pub timestamp: UnixTimestamp,
    // Sender of the message
    pub sender: Pubkey,
    // If the message is a response to another message
    pub replies_to: Pubkey,
    // Likes counter
    pub likes_count: u16,
    // Dislikes counter
    pub dislikes_count: u16,
    // Message sent
    pub msg: Vec<u8>,
}

impl Message {
    pub const SEED: &'static str = "message";

    pub fn new(
        kind: MessageType,
        timestamp: UnixTimestamp,
        msg: Vec<u8>,
        sender: Pubkey,
        replies_to: Pubkey,
    ) -> Self {
        Self {
            tag: Tag::Message,
            kind,
            timestamp,
            msg,
            sender,
            replies_to,
            likes_count: 0,
            dislikes_count: 0,
        }
    }

    pub fn find_key(
        index: u32,
        from_key: &Pubkey,
        to_key: &Pubkey,
        program_id: &Pubkey,
    ) -> (Pubkey, u8) {
        let i = index.to_string();
        let (key_1, key_2) = order_keys(from_key, to_key);
        let (message_key, bump) = Pubkey::find_program_address(
            &[
                Message::SEED.as_bytes(),
                i.as_bytes(),
                &key_1.to_bytes(),
                &key_2.to_bytes(),
            ],
            program_id,
        );
        (message_key, bump)
    }

    pub fn create_key(
        index: u32,
        from_key: &Pubkey,
        to_key: &Pubkey,
        program_id: &Pubkey,
        bump: u8,
    ) -> Pubkey {
        let (key_1, key_2) = order_keys(from_key, to_key);
        let i = index.to_string();
        let seeds = &[
            Message::SEED.as_bytes(),
            i.as_bytes(),
            &key_1.to_bytes(),
            &key_2.to_bytes(),
            &[bump],
        ];
        Pubkey::create_program_address(seeds, program_id).unwrap()
    }

    pub fn save(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    pub fn from_account_info(a: &AccountInfo) -> Result<Message, ProgramError> {
        let mut data = &a.data.borrow() as &[u8];
        if data[0] != Tag::Message as u8 && data[0] != Tag::Uninitialized as u8 {
            return Err(JabberError::DataTypeMismatch.into());
        }
        let result = Message::deserialize(&mut data)?;
        Ok(result)
    }
}

#[derive(BorshSerialize, BorshDeserialize)]
pub struct GroupThread {
    pub tag: Tag,
    // Whether to suggest the group in the app
    pub visible: bool,
    // Owner of the group (fee exempt)
    pub owner: Pubkey,
    // Time at which the message was sent
    pub last_message_time: UnixTimestamp,
    // Destination of the fees
    pub destination_wallet: Pubkey,
    // Message counter for PDA derivation
    pub msg_count: u32,
    // Fee per message
    pub lamports_per_message: u64,
    // PDA derivation bump
    pub bump: u8,
    // Whether users can post media (images, videos and audios)
    pub media_enabled: bool,
    // Whether admins only can post messages
    pub admin_only: bool,
    // IPFS hash of the group
    pub group_pic_hash: Option<String>,
    // Human readable group name
    pub group_name: String,
    // Admins of the group (fee exempt)
    pub admins: Vec<Pubkey>,
}

impl GroupThread {
    pub const SEED: &'static str = "group_thread";

    #[allow(clippy::too_many_arguments)]
    pub fn new(
        visible: bool,
        group_name: String,
        destination_wallet: Pubkey,
        lamports_per_message: u64,
        bump: u8,
        admins: Vec<Pubkey>,
        owner: Pubkey,
        media_enabled: bool,
        admin_only: bool,
        current_time: i64,
    ) -> Self {
        Self {
            tag: Tag::GroupThread,
            visible,
            group_name,
            msg_count: 0,
            destination_wallet,
            lamports_per_message,
            bump,
            admins,
            owner,
            media_enabled,
            group_pic_hash: None,
            admin_only,
            last_message_time: current_time,
        }
    }

    pub fn create_key(group_name: String, owner: Pubkey, program_id: &Pubkey, bump: u8) -> Pubkey {
        let seeds = &[
            GroupThread::SEED.as_bytes(),
            group_name.as_bytes(),
            &owner.to_bytes(),
            &[bump],
        ];
        Pubkey::create_program_address(seeds, program_id).unwrap()
    }

    pub fn find_key(group_name: String, owner: Pubkey, program_id: &Pubkey) -> (Pubkey, u8) {
        let seeds = &[
            GroupThread::SEED.as_bytes(),
            group_name.as_bytes(),
            &owner.to_bytes(),
        ];
        let (ama_thread_key, bump) = Pubkey::find_program_address(seeds, program_id);
        (ama_thread_key, bump)
    }

    pub fn save(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    pub fn increment_msg_count(&mut self, current_time: i64) {
        self.last_message_time = current_time;
        self.msg_count += 1;
    }

    pub fn from_account_info(a: &AccountInfo) -> Result<GroupThread, ProgramError> {
        let mut data = &a.data.borrow() as &[u8];
        if data[0] != Tag::GroupThread as u8 && data[0] != Tag::Uninitialized as u8 {
            return Err(JabberError::DataTypeMismatch.into());
        }
        let result = GroupThread::deserialize(&mut data)?;
        Ok(result)
    }

    pub fn is_fee_exempt(&self, sender: Pubkey, admin_index: Option<u64>) -> bool {
        if self.destination_wallet == sender {
            return true;
        }
        if let Some(idx) = admin_index {
            return self.admins[idx as usize] == sender;
        }
        false
    }

    pub fn add_admin(&mut self, admin_address: Pubkey) -> ProgramResult {
        if self.admins.len() > MAX_ADMIN_LEN {
            return Err(JabberError::MaxAdminsReached.into());
        }
        self.admins.push(admin_address);
        Ok(())
    }

    pub fn remove_admin(&mut self, admin_address: Pubkey, admin_index: usize) -> ProgramResult {
        let deleted = self.admins.remove(admin_index);
        if admin_address != deleted {
            return Err(JabberError::InvalidAdminIndex.into());
        }
        Ok(())
    }
}

// To keep track of users' groups
#[derive(BorshSerialize, BorshDeserialize, BorshSize)]
pub struct GroupThreadIndex {
    pub tag: Tag,
    // Group thread of the index
    pub group_thread_key: Pubkey,
    // Owner of the index
    pub owner: Pubkey,
    // Name of the group
    pub group_name: String,
}

impl GroupThreadIndex {
    pub const SEED: &'static str = "group_thread_index";

    pub fn new(group_name: String, group_thread_key: Pubkey, owner: Pubkey) -> Self {
        Self {
            tag: Tag::GroupThreadIndex,
            group_name,
            group_thread_key,
            owner,
        }
    }

    pub fn create_key(
        group_name: String,
        group_thread_key: Pubkey,
        owner: Pubkey,
        program_id: &Pubkey,
        bump: u8,
    ) -> Pubkey {
        let seeds = &[
            GroupThreadIndex::SEED.as_bytes(),
            group_name.as_bytes(),
            &owner.to_bytes(),
            &group_thread_key.to_bytes(),
            &[bump],
        ];
        Pubkey::create_program_address(seeds, program_id).unwrap()
    }

    pub fn find_key(
        group_name: String,
        group_thread_key: Pubkey,
        owner: Pubkey,
        program_id: &Pubkey,
    ) -> (Pubkey, u8) {
        let seeds = &[
            GroupThreadIndex::SEED.as_bytes(),
            group_name.as_bytes(),
            &owner.to_bytes(),
            &group_thread_key.to_bytes(),
        ];
        let (ama_thread_key, bump) = Pubkey::find_program_address(seeds, program_id);
        (ama_thread_key, bump)
    }

    pub fn save(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    pub fn from_account_info(a: &AccountInfo) -> Result<GroupThreadIndex, ProgramError> {
        let mut data = &a.data.borrow() as &[u8];
        if data[0] != Tag::GroupThreadIndex as u8 && data[0] != Tag::Uninitialized as u8 {
            return Err(JabberError::DataTypeMismatch.into());
        }
        let result = GroupThreadIndex::deserialize(&mut data)?;
        Ok(result)
    }
}

#[derive(BorshSerialize, BorshDeserialize)]
pub struct Subscription {
    // Pubkey of the subscriber
    pub subscriber: Pubkey,
    // Pubkey of the person the subscriber subscribed to
    pub subscribed_to: Pubkey,
}

impl Subscription {
    pub const SEED: &'static str = "subscription";

    pub fn new(subscriber: Pubkey, subscribed_to: Pubkey) -> Self {
        Self {
            subscriber,
            subscribed_to,
        }
    }

    pub fn create_key(
        subscriber: &Pubkey,
        subscribed_to: &Pubkey,
        program_id: &Pubkey,
        bump: u8,
    ) -> Pubkey {
        let seeds = &[
            Subscription::SEED.as_bytes(),
            &subscriber.to_bytes(),
            &subscribed_to.to_bytes(),
            &[bump],
        ];
        Pubkey::create_program_address(seeds, program_id).unwrap()
    }

    pub fn find_key(
        subscriber: &Pubkey,
        subscribed_to: &Pubkey,
        program_id: &Pubkey,
    ) -> (Pubkey, u8) {
        let seeds = &[
            Subscription::SEED.as_bytes(),
            &subscriber.to_bytes(),
            &subscribed_to.to_bytes(),
        ];
        Pubkey::find_program_address(seeds, program_id)
    }

    pub fn save(&self, mut dst: &mut [u8]) {
        self.serialize(&mut dst).unwrap()
    }

    pub fn from_account_info(a: &AccountInfo) -> Result<Subscription, ProgramError> {
        let mut data = &a.data.borrow() as &[u8];
        if data[0] != Tag::Subscription as u8 && data[0] != Tag::Uninitialized as u8 {
            return Err(JabberError::DataTypeMismatch.into());
        }
        let result = Subscription::deserialize(&mut data)?;
        Ok(result)
    }
}

File: js/src/state.ts
import { PublicKey, Connection } from "@solana/web3.js";
import BN from "bn.js";
import { Schema, deserializeUnchecked, deserialize } from "borsh";
import { JABBER_ID } from "./bindings";
import { orderKeys } from "./utils";

export enum Tag {
  Uninitialized = 0,
  Profile = 1,
  Thread = 2,
  Message = 3,
  Jabber = 4,
  GroupThread = 5,
  GroupThreadIndex = 6,
  Subscription = 7,
}

export enum MessageType {
  Encrypted = 0,
  Unencrypted = 1,
  EncryptedImage = 2,
  UnencryptedImage = 3,
}

export class Profile {
  tag: Tag;
  bump: number;
  pictureHash: String;
  displayDomainName: String;
  bio: String;
  lamportsPerMessage: BN;
  allowDm: boolean;
  tipsSent: number;
  tipsReceived: number;

  static schema: Schema = new Map([
    [
      Profile,
      {
        kind: "struct",
        fields: [
          ["tag", "u8"],
          ["bump", "u8"],
          ["pictureHash", "string"],
          ["displayDomainName", "string"],
          ["bio", "string"],
          ["lamportsPerMessage", "u64"],
          ["allowDm", "u8"],
          ["tipsSent", "u64"],
          ["tipsReceived", "u64"],
        ],
      },
    ],
  ]);

  constructor(obj: {
    tag: Tag;
    bump: number;
    pictureHash: String;
    displayDomainName: String;
    bio: String;
    lamportsPerMessage: BN;
    allowDm: boolean;
    tipsSent: number;
    tipsReceived: number;
  }) {
    this.tag = Tag.Profile;
    this.bump = obj.bump;
    this.pictureHash = obj.pictureHash;
    this.displayDomainName = obj.displayDomainName;
    this.bio = obj.bio;
    this.lamportsPerMessage = obj.lamportsPerMessage;
    this.allowDm = obj.allowDm;
    this.tipsSent = obj.tipsSent;
    this.tipsReceived = obj.tipsReceived;
  }

  static deserialize(data: Buffer) {
    return deserializeUnchecked(this.schema, Profile, data);
  }

  static async retrieve(connection: Connection, owner: PublicKey) {
    const [profile] = await PublicKey.findProgramAddress(
      Profile.generateSeeds(owner),
      JABBER_ID
    );

    const accountInfo = await connection.getAccountInfo(profile);

    if (!accountInfo?.data) {
      throw new Error("No profile found");
    }

    return this.deserialize(accountInfo?.data);
  }

  static generateSeeds(profileOwner: PublicKey) {
    return [Buffer.from("profile"), profileOwner.toBuffer()];
  }
}

export class Thread {
  tag: Tag;
  msgCount: number;
  user1: PublicKey;
  user2: PublicKey;
  lastMessageTime: BN;
  bump: number;

  static schema: Schema = new Map([
    [
      Thread,
      {
        kind: "struct",
        fields: [
          ["tag", "u8"],
          ["msgCount", "u32"],
          ["user1", [32]],
          ["user2", [32]],
          ["lastMessageTime", "u64"],
          ["bump", "u8"],
        ],
      },
    ],
  ]);

  constructor(obj: {
    msgCount: number;
    user1: Uint8Array;
    user2: Uint8Array;
    lastMessageTime: BN;
    bump: number;
  }) {
    this.tag = Tag.Thread;
    this.msgCount = obj.msgCount;
    this.user1 = new PublicKey(obj.user1);
    this.user2 = new PublicKey(obj.user2);
    this.lastMessageTime = obj.lastMessageTime;
    this.bump = obj.bump;
  }

  static deserialize(data: Buffer) {
    return deserialize(this.schema, Thread, data);
  }

  static generateSeeds(sender: PublicKey, receiver: PublicKey) {
    const [key1, key2] = orderKeys(sender, receiver);
    return [Buffer.from("thread"), key1.toBuffer(), key2.toBuffer()];
  }

  static async getKeys(sender: PublicKey, receiver: PublicKey) {
    const [thread] = await PublicKey.findProgramAddress(
      Thread.generateSeeds(sender, receiver),
      JABBER_ID
    );
    return thread;
  }

  static async retrieve(
    connection: Connection,
    sender: PublicKey,
    receiver: PublicKey
  ) {
    const [thread] = await PublicKey.findProgramAddress(
      Thread.generateSeeds(sender, receiver),
      JABBER_ID
    );
    const accountInfo = await connection.getAccountInfo(thread);

    if (!accountInfo?.data) {
      throw new Error("Thread not found");
    }

    return this.deserialize(accountInfo.data);
  }

  static async retrieveFromKey(connection: Connection, key: PublicKey) {
    const accountInfo = await connection.getAccountInfo(key);

    if (!accountInfo?.data) {
      throw new Error("Thread not found");
    }

    return this.deserialize(accountInfo.data);
  }
}

export class Message {
  tag: Tag;
  kind: MessageType;
  timestamp: BN;
  sender: PublicKey;
  repliesTo: PublicKey;
  likesCount: number;
  dislikesCount: number;
  msg: Uint8Array;

  static schema: Schema = new Map([
    [
      Message,
      {
        kind: "struct",
        fields: [
          ["tag", "u8"],
          ["kind", "u8"],
          ["timestamp", "u64"],
          ["sender", [32]],
          ["repliesTo", [32]],
          ["likesCount", "u16"],
          ["dislikesCount", "u16"],
          ["msg", ["u8"]],
        ],
      },
    ],
  ]);

  constructor(obj: {
    tag: Tag;
    kind: MessageType;
    timestamp: BN;
    sender: Uint8Array;
    repliesTo: Uint8Array;
    likesCount: number;
    dislikesCount: number;
    msg: Uint8Array;
  }) {
    this.tag = Tag.Message;
    this.kind = obj.kind;
    this.timestamp = obj.timestamp;
    this.sender = new PublicKey(obj.sender);
    this.repliesTo = new PublicKey(obj.repliesTo);
    this.likesCount = obj.likesCount;
    this.dislikesCount = obj.dislikesCount;
    this.msg = obj.msg;
  }

  static deserialize(data: Buffer) {
    return deserializeUnchecked(this.schema, Message, data);
  }

  static generateSeeds(
    messageCount: number,
    sender: PublicKey,
    receiver: PublicKey
  ) {
    const [key1, key2] = orderKeys(sender, receiver);
    return [
      Buffer.from("message"),
      Buffer.from(messageCount.toString()),
      key1.toBuffer(),
      key2.toBuffer(),
    ];
  }

  static async retrieveFromIndex(
    connection: Connection,
    index: number,
    receiver: PublicKey,
    sender: PublicKey
  ) {
    const [messageAccount] = await PublicKey.findProgramAddress(
      this.generateSeeds(index, sender, receiver),
      JABBER_ID
    );
    const accountInfo = await connection.getAccountInfo(messageAccount);
    if (!accountInfo?.data) {
      throw new Error("Invalid message info");
    }
    return this.deserialize(accountInfo.data);
  }

  static async retrieveFromThread(
    connection: Connection,
    sender: PublicKey,
    receiver: PublicKey
  ) {
    const thread = await Thread.retrieve(connection, sender, receiver);
    let messageAccounts: PublicKey[] = [];
    for (let i = 0; i < thread.msgCount; i++) {
      const [acc] = await PublicKey.findProgramAddress(
        this.generateSeeds(i, sender, receiver),
        JABBER_ID
      );
      messageAccounts.push(acc);
    }
    const accountInfos = await connection.getMultipleAccountsInfo(
      messageAccounts
    );
    return accountInfos.map((info, i) =>
      info?.data
        ? { message: this.deserialize(info?.data), address: messageAccounts[i] }
        : undefined
    );
  }
}

export class GroupThread {
  tag: Tag;
  bump: number;
  visible: boolean;
  owner: PublicKey;
  lastMessageTime: BN;
  destinationWallet: PublicKey;
  msgCount: number;
  lamportsPerMessage: BN;
  mediaEnabled: boolean;
  adminOnly: boolean;
  groupPicHash: String;
  groupName: string;
  admins: PublicKey[];

  static schema: Schema = new Map([
    [
      GroupThread,
      {
        kind: "struct",
        fields: [
          ["tag", "u8"],
          ["bump", "u8"],
          ["visible", "u8"],
          ["owner", [32]],
          ["lastMessageTime", "u64"],
          ["destinationWallet", [32]],
          ["msgCount", "u32"],
          ["lamportsPerMessage", "u64"],
          ["mediaEnabled", "u8"],
          ["adminOnly", "u8"],
          ["groupPicHash", "string"],
          ["groupName", "string"],
          ["admins", [[32]]],
        ],
      },
    ],
  ]);

  constructor(obj: {
    tag: Tag;
    bump: number;
    visible: boolean;
    owner: PublicKey;
    lastMessageTime: BN;
    destinationWallet: PublicKey;
    msgCount: number;
    lamportsPerMessage: BN;
    mediaEnabled: boolean;
    adminOnly: boolean;
    groupPicHash: String;
    groupName: string;
    admins: PublicKey[];
  }) {
    this.tag = Tag.GroupThread;
    this.bump = obj.bump;
    this.visible = obj.visible;
    this.owner = new PublicKey(obj.owner);
    this.lastMessageTime = obj.lastMessageTime;
    this.destinationWallet = new PublicKey(obj.destinationWallet);
    this.msgCount = obj.msgCount;
    this.lamportsPerMessage = obj.lamportsPerMessage;
    this.mediaEnabled = !!obj.mediaEnabled;
    this.adminOnly = !!obj.adminOnly;
    this.groupPicHash = obj.groupPicHash;
    this.groupName = obj.groupName;
    this.admins = obj.admins.map((e) => new PublicKey(e));
  }

  static deserialize(data: Buffer) {
    return deserializeUnchecked(this.schema, GroupThread, data);
  }

  static generateSeeds(groupName: string, owner: PublicKey) {
    return [
      Buffer.from("group_thread"),
      Buffer.from(groupName),
      owner.toBuffer(),
    ];
  }

  static async getKey(groupName: string, owner: PublicKey) {
    const [groupThread] = await PublicKey.findProgramAddress(
      GroupThread.generateSeeds(groupName, owner),
      JABBER_ID
    );
    return groupThread;
  }

  static async retrieve(
    connection: Connection,
    groupName: string,
    owner: PublicKey
  ) {
    const groupThreadKey = await GroupThread.getKey(groupName, owner);

    const accountInfo = await connection.getAccountInfo(groupThreadKey);

    if (!accountInfo?.data) {
      throw new Error("Group thread not found");
    }

    return this.deserialize(accountInfo.data);
  }

  static async retrieveFromKey(connection: Connection, key: PublicKey) {
    const accountInfo = await connection.getAccountInfo(key);

    if (!accountInfo?.data) {
      throw new Error("Group thread not found");
    }

    return this.deserialize(accountInfo.data);
  }
}

export class GroupThreadIndex {
  tag: number;
  groupThreadKey: Uint8Array;
  owner: Uint8Array;
  groupName: string;

  static schema: Schema = new Map([
    [
      GroupThreadIndex,
      {
        kind: "struct",
        fields: [
          ["tag", "u8"],
          ["groupThreadKey", [32]],
          ["owner", [32]],
          ["groupName", "string"],
        ],
      },
    ],
  ]);

  constructor(obj: {
    groupName: string;
    groupThreadKey: Uint8Array;
    owner: Uint8Array;
  }) {
    this.tag = Tag.GroupThreadIndex;
    this.groupName = obj.groupName;
    this.groupThreadKey = obj.groupThreadKey;
    this.owner = obj.owner;
  }

  static deserialize(data: Buffer) {
    return deserializeUnchecked(this.schema, GroupThreadIndex, data);
  }

  static generateSeeds(
    groupName: string,
    owner: PublicKey,
    groupThreadKey: PublicKey
  ) {
    return [
      Buffer.from("group_thread_index"),
      Buffer.from(groupName),
      owner.toBuffer(),
      groupThreadKey.toBuffer(),
    ];
  }

  static async getKey(
    groupName: string,
    owner: PublicKey,
    groupThreadKey: PublicKey
  ) {
    const [groupThreadIndex] = await PublicKey.findProgramAddress(
      GroupThreadIndex.generateSeeds(groupName, owner, groupThreadKey),
      JABBER_ID
    );
    return groupThreadIndex;
  }

  static async retrieve(
    connection: Connection,
    groupName: string,
    owner: PublicKey,
    groupThreadKey: PublicKey
  ) {
    const groupThreadIndexKey = await GroupThreadIndex.getKey(
      groupName,
      owner,
      groupThreadKey
    );

    const accountInfo = await connection.getAccountInfo(groupThreadIndexKey);

    if (!accountInfo?.data) {
      throw new Error("Group index not found");
    }

    return this.deserialize(accountInfo.data);
  }

  static async retrieveFromKey(connection: Connection, key: PublicKey) {
    const accountInfo = await connection.getAccountInfo(key);

    if (!accountInfo?.data) {
      throw new Error("Group index not found");
    }

    return this.deserialize(accountInfo.data);
  }
}

export class Subscription {
  tag: number;
  subscriber: Uint8Array;
  subscribedTo: Uint8Array;

  static schema: Schema = new Map([
    [
      Subscription,
      {
        kind: "struct",
        fields: [
          ["tag", "u8"],
          ["subscriber", [32]],
          ["subscribedTo", [32]],
        ],
      },
    ],
  ]);

  constructor(obj: { subscriber: Uint8Array; subscribedTo: Uint8Array }) {
    this.tag = Tag.Subscription;
    this.subscriber = obj.subscriber;
    this.subscribedTo = obj.subscribedTo;
  }

  static deserialize(data: Buffer) {
    return deserializeUnchecked(this.schema, Subscription, data);
  }

  static generateSeeds(subscriber: PublicKey, subscribedTo: PublicKey) {
    return [
      Buffer.from("subscription"),
      subscriber.toBuffer(),
      subscribedTo.toBuffer(),
    ];
  }

  static async getKey(subscriber: PublicKey, subscribedTo: PublicKey) {
    const [subscriptionKey] = await PublicKey.findProgramAddress(
      Subscription.generateSeeds(subscriber, subscribedTo),
      JABBER_ID
    );
    return subscriptionKey;
  }
}
